src/aggregators/prioritytable.jl

```diff
@@ -312,3 +318,155 @@ function sample(pt::PriorityTable, priorities, rng = DEFAULT_RNG)
+    groups::Vector{PriorityGroup{T, Vector{F}}}
+    pidtogroup::Vector{Tuple{F, F}}
+    times::Vector{T}
+    timegrouper::TimeGrouper{T}
+    minbin::F
+    steps::F # TODO: For adaptive rebuilding. 
+    maxtime::T
+end
+
+# Construct the time table with the default optimal bin width and number of bins. 
+# DEFAULT NUMBINS: 20 * √length(times)
+# DEFAULT BINWIDTH: 16 / sum(propensities)
+function PriorityTimeTable(times::AbstractVector, mintime, timestep)
+    numbins = floor(Int64, 20 * sqrt(length(times)))
+    maxtime = mintime + numbins * timestep
+
+    pidtype = typeof(numbins)
+    ptype = eltype(times)
+    groups = Vector{PriorityGroup{ptype, Vector{pidtype}}}()
+    pidtogroup = Vector{Tuple{Int, Int}}(undef, length(times))
+
+    ttgdata = TimeGrouper{ptype}(mintime, timestep)
+    # Create the groups, [t_min, t_min + τ), [t_min + τ, t_min + 2τ)...
+    for i in 1:numbins
+        push!(groups, PriorityGroup{pidtype}(mintime + i * timestep))
+    end
+
+    ptt = PriorityTimeTable(
+        groups, pidtogroup, times, ttgdata, zero(pidtype), zero(pidtype), maxtime)
+    # Insert priority ids into the groups
+    for (pid, time) in enumerate(times)
+        if time > maxtime
+            pidtogroup[pid] = (0, 0)
+            continue
+        end
+        insert!(ptt, pid, time)
+    end
+
+    ptt.minbin = findfirst(g -> g.numpids > (0), groups)
+    ptt.minbin === nothing && (ptt.minbin = 0)
+    ptt
+end
+
+# Rebuild the table when there are no more reaction times within the current
+# time window. 
+function rebuild!(ptt::PriorityTimeTable{T, F}, mintime, timestep) where {T, F}
+    @unpack pidtogroup, groups, times, timegrouper = ptt
+    fill!(pidtogroup, (zero(F), zero(F)))
+
+    numbins = length(groups)
+    ptt.maxtime = mintime + numbins * timestep
+    timegrouper.mintime = mintime
+    timegrouper.timestep = timestep
+
+    groupmaxtime = mintime
+    for group in groups
+        group.numpids = zero(F)
+        groupmaxtime += timestep
+        group.maxpriority = groupmaxtime
+    end
+
+    # Reinsert the times into the groups. 
+    for (id, time) in enumerate(times)
+        time > ptt.maxtime && continue
+        insert!(ptt, id, time)
+    end
+    ptt.minbin = findfirst(g -> g.numpids > (0), groups)
+    ptt.minbin === nothing && (ptt.minbin = 0)
+    ptt.steps = 0
+end
+
+# Get the reaction with the earliest timestep.
+function getfirst(ptt::PriorityTimeTable)
+    @unpack groups, times, minbin = ptt
+    minbin == 0 && return (0, 0)
+
+    while groups[minbin].numpids == 0
+        minbin += 1
+        if minbin > length(groups)
+            return (0, 0)
+        end
+    end
+
+    ptt.minbin = minbin
+    ptt.steps += 1
+    min_time = typemax(eltype(times))
+    min_idx = 0
+    for i in 1:(groups[minbin].numpids)
+        pid = groups[minbin].pids[i]
+        times[pid] < min_time && begin
+            min_time = times[pid]
+            min_idx = pid
+        end
+    end
```

Comment:
```suggestion
    for pid in groups[minbin].pids
        times[pid] < min_time && begin
            min_time = times[pid]
            min_idx = pid
        end
    end
```

Reply:
A bit uncertain if this works - since the number of active PIDs may be fewer than the total number of PIDs you could be accessing PIDs that are no longer in the bin if you iterate over all of them.